File: Example/ObjC/Bulletin/PetValidationBulletinItem.m
/**
 *  BulletinBoard
 *  Copyright (c) 2017 - present Alexis Aubry. Licensed under the MIT license.
 */

#import "PetValidationBulletinItem.h"
#import "BulletinDataSource.h"
#import "ImageCollectionViewCell.h"
#import "SelectionFeedbackGenerator.h"
#import "SuccessFeedbackGenerator.h"

@interface PetValidationBulletinItem ()

@property (nonatomic, strong) CollectionDataSource *dataSource;
@property (nonatomic, strong) SelectionFeedbackGenerator *selectionFeedbackGenerator;
@property (nonatomic, strong) SuccessFeedbackGenerator *successFeedbackGenerator;
@property (nonatomic, strong, nullable) UICollectionView *collectionView;

@end

@interface PetValidationBulletinItem (CollectionView) <UICollectionViewDelegateFlowLayout, UICollectionViewDataSource>

@end

@implementation PetValidationBulletinItem

- (instancetype)initWithDataSource:(CollectionDataSource *)data
{
    self = [super initWithTitle:@"Choose your Favorite"];
    if (self) {
        self.dataSource = data;
        self.selectionFeedbackGenerator = [[SelectionFeedbackGenerator alloc] init];
        self.successFeedbackGenerator = [[SuccessFeedbackGenerator alloc] init];
        self.collectionView = nil;
        self.descriptionText = [[NSString alloc] initWithFormat:@"You chose %@ as your favorite animal type. Here are a few examples of posts in this category.", [[data pluralizedPetName] lowercaseString]];
        self.actionButtonTitle = @"Validate";
        self.alternativeButtonTitle = @"Change";
    }
    return self;
}

- (NSArray<UIView *> *)makeViewsUnderDescriptionWithInterfaceBuilder:(BLTNInterfaceBuilder *)interfaceBuilder
{
    UICollectionViewFlowLayout *layout = [[UICollectionViewFlowLayout alloc] init];
    layout.scrollDirection = UICollectionViewScrollDirectionVertical;
    layout.minimumInteritemSpacing = 1;

    UICollectionView *collectionView = [[UICollectionView alloc] initWithFrame:CGRectZero collectionViewLayout:layout];
    collectionView.backgroundColor = [UIColor whiteColor];

    BLTNContainerView *collectionWrapper = [interfaceBuilder wrapView:collectionView width:NULL height:@256 position:BLTNViewPositionPinnedToEdges];

    self.collectionView = collectionView;
    return @[collectionWrapper];

}

- (void)setUp
{
    [self.collectionView registerClass:[ImageCollectionViewCell class] forCellWithReuseIdentifier:@"cell"];
    self.collectionView.dataSource = self;
    self.collectionView.delegate = self;
    [super setUp];
}

- (void)tearDown
{
    self.collectionView.dataSource = nil;
    self.collectionView.delegate = nil;
}

#pragma mark - Touch Events

- (void)actionButtonTappedWithSender:(UIButton *)sender
{
    // Play selection haptic feedback
    [self.selectionFeedbackGenerator prepare];
    [self.selectionFeedbackGenerator selectionChanged];

    // Display the loading indicator
    [self.manager displayActivityIndicatorWithColor:[UIColor blackColor]];

    // Wait for a "task" to complete before displaying the next item

    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(2 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        // Play success haptic feedback
        [self.successFeedbackGenerator prepare];
        [self.successFeedbackGenerator notifySuccess];

        // Display next item
        self.nextItem = [BulletinDataSource makeCompletionPage];
        [self.manager displayNextItem];
    });
}

-(void)alternativeButtonTappedWithSender:(UIButton *)sender
{
    // Play selection haptic feedback
    [self.selectionFeedbackGenerator prepare];
    [self.selectionFeedbackGenerator selectionChanged];

    // Display previous item
    [self.manager popItem];
}

#pragma mark - Collection View

- (NSInteger)numberOfSectionsInCollectionView:(UICollectionView *)collectionView
{
    return 1;
}

- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section
{
    return 9;
}

- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath
{
    ImageCollectionViewCell *cell = (ImageCollectionViewCell *)[collectionView dequeueReusableCellWithReuseIdentifier:@"cell" forIndexPath:indexPath];
    cell.imageView.image = [self.dataSource imageAtIndex:indexPath.row];
    cell.imageView.contentMode = UIViewContentModeScaleAspectFill;
    cell.imageView.clipsToBounds = YES;
    return cell;
}

- (CGSize)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout sizeForItemAtIndexPath:(NSIndexPath *)indexPath
{
    CGFloat squareSideLength = (collectionView.frame.size.width / 3) - 3;
    return CGSizeMake(squareSideLength, squareSideLength);
}

@end

File: Example/CustomBulletins/PetValidationBulletinItem.swift
/**
 *  BulletinBoard
 *  Copyright (c) 2017 - present Alexis Aubry. Licensed under the MIT license.
 */

import UIKit
import BLTNBoard

/**
 * A bulletin page that allows the user to validate its selection.
 *
 * This item demonstrates popping to the previous item, and including a collection view inside the page.
 */

@objc public class PetValidationBLTNItem: FeedbackPageBLTNItem {

    let dataSource: CollectionDataSource
    let animalType: String
    let validationHandler: (BLTNItem) -> Void

    let selectionFeedbackGenerator = SelectionFeedbackGenerator()
    let successFeedbackGenerator = SuccessFeedbackGenerator()

    init(dataSource: CollectionDataSource, animalType: String, validationHandler: @escaping (BLTNItem) -> Void) {
        self.dataSource = dataSource
        self.animalType = animalType
        self.validationHandler = validationHandler
        super.init(title: "Choose your Favorite")

        isDismissable = false
        descriptionText = "You chose \(animalType) as your favorite animal type. Here are a few examples of posts in this category."
        actionButtonTitle = "Validate"
        alternativeButtonTitle = "Change"

    }

    // MARK: - Interface

    var collectionView: UICollectionView?

    override public func makeViewsUnderDescription(with interfaceBuilder: BLTNInterfaceBuilder) -> [UIView]? {

        let flowLayout = UICollectionViewFlowLayout()
        flowLayout.scrollDirection = .vertical
        flowLayout.minimumInteritemSpacing = 1

        let collectionView = UICollectionView(frame: .zero, collectionViewLayout: flowLayout)
        collectionView.backgroundColor = .white

        let collectionWrapper = interfaceBuilder.wrapView(collectionView, width: nil, height: 256, position: .pinnedToEdges)

        self.collectionView = collectionView
        collectionView.register(ImageCollectionViewCell.self, forCellWithReuseIdentifier: "cell")
        collectionView.dataSource = self
        collectionView.delegate = self

        return [collectionWrapper]

    }

    override public func tearDown() {
        super.tearDown()
        collectionView?.dataSource = nil
        collectionView?.delegate = nil
    }

    // MARK: - Touch Events

    override public func actionButtonTapped(sender: UIButton) {

        // > Play Haptic Feedback

        selectionFeedbackGenerator.prepare()
        selectionFeedbackGenerator.selectionChanged()

        // > Display the loading indicator

        manager?.displayActivityIndicator()

        // > Wait for a "task" to complete before displaying the next item

        let delay = DispatchTime.now() + .seconds(2)

        DispatchQueue.main.asyncAfter(deadline: delay) {
            // Play success haptic feedback
            self.successFeedbackGenerator.prepare()
            self.successFeedbackGenerator.success()

            // Display next item
            self.validationHandler(self)
        }
    }

    public override func alternativeButtonTapped(sender: UIButton) {

        // Play selection haptic feedback

        selectionFeedbackGenerator.prepare()
        selectionFeedbackGenerator.selectionChanged()

        // Display previous item

        manager?.popItem()

    }

}

// MARK: - Collection View

extension PetValidationBLTNItem: UICollectionViewDataSource, UICollectionViewDelegateFlowLayout {

    public func numberOfSections(in collectionView: UICollectionView) -> Int {
        return 1
    }

    public func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        return 9
    }

    public func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {

        let cell = collectionView.dequeueReusableCell(withReuseIdentifier: "cell", for: indexPath) as! ImageCollectionViewCell
        cell.imageView.image = dataSource.image(at: indexPath.row)
        cell.imageView.contentMode = .scaleAspectFill
        cell.imageView.clipsToBounds = true

        return cell

    }

    public func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, sizeForItemAt indexPath: IndexPath) -> CGSize {

        let squareSideLength = (collectionView.frame.width / 3) - 3
        return CGSize(width: squareSideLength, height: squareSideLength)

    }

}
